lib/logstash/inputs/gelf.rb

```diff
@@ -68,36 +76,135 @@ def initialize(params)
   public
   def register
     require 'gelfd'
+    @tcp = nil
+    @udp = nil
+    if @port_tcp == nil
+      @port_tcp = @port
+    end
+    if @port_udp == nil
+      @port_udp = @port
+    end  
   end # def register
 
   public
   def run(output_queue)
     begin
-      # udp server
-      udp_listener(output_queue)
+      if @use_tcp
+        tcp_thr = Thread.new(output_queue) do |output_queue|
+          tcp_listener(output_queue)
+        end
+      end
+      if @use_udp
+        udp_thr = Thread.new(output_queue) do |output_queue|
+          udp_listener(output_queue)
+        end
+      end
     rescue => e
       unless stop?
         @logger.warn("gelf listener died", :exception => e, :backtrace => e.backtrace)
         Stud.stoppable_sleep(RECONNECT_BACKOFF_SLEEP) { stop? }
         retry unless stop?
       end
     end # begin
+    if @use_tcp
+      tcp_thr.join
+    end
+    if @use_udp
+      udp_thr.join
+    end
   end # def run
 
   public
   def stop
-    @udp.close
+    if @use_tcp && @use_udp
+      return @udp.close && tcp.close
+    elseif @use_tcp
```

Comment:
elseif should be elsif and same for below

Reply:
Done